### PR TITLE
Implement conditional stripping of type descriptors, protocols and protocol conformances via !llvm.used.conditional

### PR DESCRIPTION
**Pull request by kubamracek**
A new LLVM IR affordance that allows expressing conditions under which globals
can be removed/dropped (even when marked with @llvm.used) is being discussed at:

- <https://reviews.llvm.org/D104496>
- <https://lists.llvm.org/pipermail/llvm-dev/2021-September/152656.html>

This is a preliminary implementation that marks runtime lookup records (namely
protocol records, type descriptors records and protocol conformance records)
with the !llvm.used.conditional descriptors. That allows link-time / LTO-time
removal of these records (by GlobalDCE) based on whether they're actually used
within the linkage unit. Effectively, this allows libraries that have a limited
and known set of clients, to be optimized against the client at LTO time, and
significantly reduce the code size of that library.

Parts of the implementation:
- New -conditional-runtime-records frontend flag to enable using !llvm.used.conditional
- IRGen code that emits these records can now emit these either as a single contiguous
  array (asContiguousArray = true, the old way), which is used for JIT mode, or
  as indivial globals (asContiguousArray = false), which is necessary for the
  !llvm.used.conditional stripping to work.
- When records are emitted as individual globals, they have new names of
  "\01l_protocol_" + mangled name of the protocol descriptor, and similarly for
  other records.
- Fixed existing tests to account for individual records instead of a single array
- Added an IR level test, and an end-to-end execution test to demonstrate that
  the !llvm.used.conditional-based stripping actually works.

### CODE REVIEW

**Comment by kubamracek** [comment]
@swift-ci please test

File: lib/IRGen/GenDecl.cpp [hunk 1]
```diff
@@ -3649,34 +3651,62 @@ IRGenModule::emitDirectRelativeReference(llvm::Constant *target,
...
+      IGM.Module.getOrInsertNamedMetadata("llvm.used.conditional");
+  usedConditional->addOperand(
+      llvm::MDNode::get(IGM.Module.getContext(), metadata));
+}
 
-    record.finishAndAddTo(recordsArray);
-  }
+static void
+appendLLVMUsedConditionalEntry(IRGenModule &IGM, llvm::GlobalVariable *var,
+                               const ConformanceDescription &record) {
```
Comment (eeckstein) [inline lib/IRGen/GenDecl.cpp:3675]: Can you add a comment saying that `var` depends on record.protocol and record.type

File: lib/IRGen/GenDecl.cpp [hunk 2]
```diff
@@ -3697,12 +3727,60 @@ llvm::Constant *IRGenModule::emitSwiftProtocols() {
...
+  }
+
+  // In non-JIT mode, emit the protocol records as individual globals.
+  for (auto *protocol : SwiftProtocols) {
+    auto entity = LinkEntity::forProtocolDescriptor(protocol);
+    auto link = LinkInfo::get(*this, entity, NotForDefinition);
+    auto var = new llvm::GlobalVariable(
+        Module, ProtocolRecordTy, /*isConstant*/ true,
+        llvm::GlobalValue::PrivateLinkage, /*initializer*/ nullptr,
+        "\x01l_protocol_" + link.getName());
```
Comment (eeckstein) [inline lib/IRGen/GenDecl.cpp:3763]: That naming convention breaks the ability to demangle the symbol. Also, it adds a lot of symbol name string data.
I'd prefer to add a mangling scheme for these variables.

File: lib/IRGen/GenDecl.cpp [hunk 3]
```diff
@@ -3697,12 +3727,60 @@ llvm::Constant *IRGenModule::emitSwiftProtocols() {
...
+    llvm::Constant *recordFields[] = {relativeAddr};
+    auto record = llvm::ConstantStruct::get(ProtocolRecordTy, recordFields);
+    var->setInitializer(record);
+
+    var->setSection(sectionName);
+    var->setAlignment(llvm::MaybeAlign(4));
+    disableAddressSanitizer(*this, var);
+    addUsedGlobal(var);
+
+    if (IRGen.Opts.ConditionalRuntimeRecords) {
```
Comment (eeckstein) [inline lib/IRGen/GenDecl.cpp:3778]: Can you add a comment describing what depends on what (it's not obvious to me).

File: lib/IRGen/GenDecl.cpp [hunk 4]
```diff
@@ -3758,17 +3817,70 @@ llvm::Constant *IRGenModule::emitProtocolConformances() {
...
+  // In non-JIT mode, emit the protocol conformance records as individual
+  // globals.
+  for (const auto &record : ProtocolConformances) {
+    auto entity =
+        LinkEntity::forProtocolConformanceDescriptor(record.conformance);
+    auto link = LinkInfo::get(*this, entity, NotForDefinition);
+    auto var = new llvm::GlobalVariable(
+        Module, RelativeAddressTy, /*isConstant*/ true,
+        llvm::GlobalValue::PrivateLinkage, /*initializer*/ nullptr,
+        "\x01l_protocol_conformance_" + link.getName());
```
Comment (eeckstein) [inline lib/IRGen/GenDecl.cpp:3855]: Same here: it's better to add a mangled name

File: lib/IRGen/GenDecl.cpp [hunk 5]
```diff
@@ -3758,17 +3817,70 @@ llvm::Constant *IRGenModule::emitProtocolConformances() {
...
+    llvm::Constant *relativeAddr =
+        emitDirectRelativeReference(descriptorRef.getValue(), var, {});
+    var->setInitializer(relativeAddr);
+
+    var->setSection(sectionName);
+    var->setAlignment(llvm::MaybeAlign(4));
+    disableAddressSanitizer(*this, var);
+    addUsedGlobal(var);
+
+    if (IRGen.Opts.ConditionalRuntimeRecords) {
```
Comment (eeckstein) [inline lib/IRGen/GenDecl.cpp:3867]: Again: please add a comment describing what depends on what

File: lib/IRGen/GenDecl.cpp [hunk 6]
```diff
@@ -3788,54 +3900,77 @@ llvm::Constant *IRGenModule::emitTypeMetadataRecords() {
...
   }
 
-  auto initializer = llvm::ConstantArray::get(arrayTy, elts);
+  // In non-JIT mode, emit the type records as individual globals.
+  for (auto type : RuntimeResolvableTypes) {
+    auto ref = getTypeEntityReference(type);
+    auto var = new llvm::GlobalVariable(
+        Module, TypeMetadataRecordTy, /*isConstant*/ true,
+        llvm::GlobalValue::PrivateLinkage, /*initializer*/ nullptr,
+        "\x01l_type_metadata_" + ref.getValue()->getName());
```
Comment (eeckstein) [inline lib/IRGen/GenDecl.cpp:3958]: again: mangling

File: lib/IRGen/GenDecl.cpp [hunk 7]
```diff
@@ -3788,54 +3900,77 @@ llvm::Constant *IRGenModule::emitTypeMetadataRecords() {
...
-  
-  addUsedGlobal(var);
-  return var;
+    var->setSection(sectionName);
+    var->setAlignment(llvm::MaybeAlign(4));
+    disableAddressSanitizer(*this, var);
+    addUsedGlobal(var);
+
+    if (IRGen.Opts.ConditionalRuntimeRecords) {
+      appendLLVMUsedConditionalEntry(*this, var, ref.getValue());
```
Comment (eeckstein) [inline lib/IRGen/GenDecl.cpp:3969]: again: comment

**Review by eeckstein — COMMENTED** [review]
basically LGTM.
See my comments

**Comment by kubamracek** [comment]
@swift-ci please test

**Comment by kubamracek** [comment]
@swift-ci please test

**Comment by kubamracek** [comment]
@swift-ci please test

**Comment by kubamracek** [comment]
@swift-ci please test Windows platform

**Review by eeckstein — APPROVED** [review]
LGTM!

@aschwaighofer can to please take a look, too?

**Review by aschwaighofer — APPROVED** [review]
lgtm

**Comment by kubamracek** [comment]
Thanks for the review!

**Pull request merged**